include/revng/Support/GraphAlgorithms.h

```diff
@@ -518,26 +541,844 @@ nodesBetweenNewImpl(GraphT Source,
+template<class NodeT>
+bool equal(SmallSetVector<NodeT> &First, SmallSetVector<NodeT> &Second) {
+  return llvm::set_is_subset(First, Second)
+         and llvm::set_is_subset(Second, First);
+}
+
+template<class NodeT>
+bool simplifyRegionsStep(llvm::SmallVector<SmallSetVector<NodeT>> &R) {
+  for (auto RegionIt1 = R.begin(); RegionIt1 != R.end(); RegionIt1++) {
+    for (auto RegionIt2 = std::next(RegionIt1); RegionIt2 != R.end();
+         RegionIt2++) {
+      bool Intersects = intersect(*RegionIt1, *RegionIt2);
+      bool IsIncluded = subset(*RegionIt1, *RegionIt2);
+      bool IsIncludedReverse = subset(*RegionIt2, *RegionIt1);
+      bool AreEquivalent = equal(*RegionIt1, *RegionIt2);
+      if (Intersects
+          and (((!IsIncluded) and (!IsIncludedReverse)) or AreEquivalent)) {
+        (*RegionIt1).insert((*RegionIt2).begin(), (*RegionIt2).end());
+        R.erase(RegionIt2);
+        return true;
+      }
+    }
+  }
+
+  return false;
+}
+
+template<class NodeT>
+void simplifyRegions(llvm::SmallVector<SmallSetVector<NodeT>> &Rs) {
+  bool Changes = true;
+  while (Changes) {
+    Changes = simplifyRegionsStep(Rs);
+  }
+}
+
+// Reorder the vector containing the regions so that they are in increasing size
+// order.
+template<class NodeT>
+void sortRegions(llvm::SmallVector<SmallSetVector<NodeT>> &Rs) {
+  std::sort(Rs.begin(),
+            Rs.end(),
+            [](SmallSetVector<NodeT> &First, SmallSetVector<NodeT> &Second) {
+              return First.size() < Second.size();
+            });
+}
+
+template<class GraphT,
+         class GT = llvm::GraphTraits<GraphT>,
+         typename NodeRef = typename GT::NodeRef>
+SmallDenseMap<NodeRef, size_t> computeDistanceFromEntry(GraphT Source) {
+  SmallDenseMap<NodeRef, size_t> ShortestPathFromEntry;
+
+  using SetType = llvm::bf_iterator_default_set<NodeRef>;
+  using bf_iterator = llvm::bf_iterator<GraphT, SetType, GT>;
+  auto BFSIt = bf_iterator::begin(Source);
+  auto BFSEnd = bf_iterator::end(Source);
+
+  for (; BFSIt != BFSEnd; ++BFSIt) {
+    NodeRef Block = *BFSIt;
+    size_t Depth = BFSIt.getLevel();
+
+    // Obtain the insertion iterator for the `Depth` block element.
+    auto ShortestIt = ShortestPathFromEntry.insert({ Block, Depth });
+
+    // If we already had in the map an entry for the current block, we need to
+    // assert that the previously found value for the `Depth` is less or equal
+    // of the `Depth` we are inserting.
+    if (ShortestIt.second == false) {
+      revng_assert(ShortestIt.first->second <= Depth);
+    }
+  }
+
+  return ShortestPathFromEntry;
+}
+
+template<class NodeT>
+bool setContains(SmallSetVector<NodeT> &Set, NodeT &Element) {
+  return Set.contains(Element);
+}
+
+template<class GraphT, class GT = llvm::GraphTraits<GraphT>>
+auto child_range(GraphT Block) {
+  return llvm::make_range(GT::child_begin(Block), GT::child_end(Block));
+}
+
+template<class GraphT>
+auto successor_range(GraphT Block) {
+  return child_range(Block);
+}
```

Comment:
- I have both just to fold code, `child_range` accepts the template parameter, so that I can call it both for `successor_range` and `predecessor_range` just with a different template `GT` parameter to do the right thing and call `llvm::make_range` once. I can of course duplicate the call to `llvm::make_range` in both `successor_range` and `predecessor_range`
- I removed the `_range` suffix, but unfortunately I could not simply use the `successors` and `predecessors` name due to conflict with the `successors` and `predecessors` defined for `llvm::BasicBlock` in `llvm/include/llvm/IR/CFG.h`. I used `blockSuccessors` and the like, otherwise I can wrap inside a namespace I guess. I made this changes in bd8b633422434d6d23731090767b287409f236f1 to avoid changing references from previous comments, this will be squashed OC.

Reply:
1. ACK, but `successor_range` is exactly the same as `child_range` without specifying the default template argument. Collapse them.
2. Uhm, not a fan of `blockSuccessors`. These work for all the guys implementing graph traits. Maybe `graph_successors`?

Reply:
1. Ah ok right, I did them in the previous way to avoid confusion on successor/predecessor, but after thinking about it collasping them seems reasonable, done in 0ff3e6296c9aecc1bd81bf7c2532dfcbabd8fdfd.
2. Used `graph_successors` as suggested in 0ff3e6296c9aecc1bd81bf7c2532dfcbabd8fdfd.